### PR TITLE
Add custom prefix configuration to `DerivedBuildCodeFormatter`

### PR DESCRIPTION
**Pull request by iangmaia**
## What does it do?

This PR adds to the class `DerivedBuildCodeFormatter` the possibility to configure a versioning prefix instead of always defaulting to `1`.

## Checklist before requesting a review

- [x] Run `bundle exec rubocop` to test for code style violations and recommendations.
- [x] Add Unit Tests (aka `specs/*_spec.rb`) if applicable.
- [x] Run `bundle exec rspec` to run the whole test suite and ensure all your tests pass.
- [x] Make sure you added an entry in [the `CHANGELOG.md` file](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/CHANGELOG.md#trunk) to describe your changes under the appropriate existing `###` subsection of the existing `## Trunk` section.
- [x] If applicable, add an entry in [the `MIGRATION.md` file](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/MIGRATION.md) to describe how the changes will affect the migration from the previous major version and what the clients will need to change and consider.

### CODE REVIEW

File: lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb [hunk 1]
```diff
@@ -19,15 +28,55 @@ class DerivedBuildCodeFormatter
...
             patch: version.patch,
             build_number: version.build_number
           )
+
+          # Only trim leading zeros when prefix is empty (to avoid removing explicit "0" prefix)
+          if @prefix.empty?
+            result.gsub(/^0+/, '')
+          else
+            result
+          end
```
Comment (AliSoftware) [inline lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb]: I'm actually not sure how the Google Play API would react to submitting a build with a `versionCode` that has a leading zero 🤔

File: lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb [hunk 2]
```diff
@@ -19,15 +28,55 @@ class DerivedBuildCodeFormatter
...
+          begin
+            prefix_int = Integer(prefix_str)
+          rescue ArgumentError
+            UI.user_error!("Prefix must be an integer digit (0-9) or empty string, got: '#{prefix_str}'")
+          end
+
+          # Check if it's within valid range (0-9)
+          return if prefix_int.between?(0, 9)
+
+          UI.user_error!("Prefix must be a single digit (0-9), got: #{prefix_int}")
```
Comment (AliSoftware) [inline lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb]: I'm not sure how it'd be technically possible to get to that point in the code and fail that test, given we already ensured that `prefix_str` would not be a string longer than 1 and would be a valid `Integer` 😅 🙃

File: lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb (hunk 1 above)
Comment (iangmaia) [reply lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb]: iinm, it needs to be an integer in the API so I'd guess they would just be ignored? 🤔 But also considering iOS, perhaps it doesn't make a lot of sense to allow for 0 to be used as prefix (or maybe we should treat it the same as `''`).

Comment (AliSoftware) [reply lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb]: Yeah I think doing the `gsub` unconditionally should be fine.

File: lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb (hunk 2 above)
Comment (iangmaia) [reply lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb]: Indeed 😅 Simplified on 49edcb6621f6918a1159a175ad8ad706ae8be90d

File: lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb [hunk 3]
```diff
@@ -6,9 +6,18 @@ module Versioning
       # The `DerivedBuildCodeFormatter` class is a specialized build code formatter for derived build codes.
       # It takes in an AppVersion object and derives a build code from it.
       class DerivedBuildCodeFormatter
+        # Initialize the formatter with a configurable prefix.
+        #
+        # @param [String] prefix The prefix to use for the build code. Must be a single digit (0-9). Defaults to '1' for backward compatibility.
```
Comment (AliSoftware) [inline lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb]:
```suggestion
        # @param [String] prefix The prefix to use for the build code. Must be a single digit (0-9), or the empty string. Defaults to '1' for backward compatibility.
```

File: lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb [hunk 4]
```diff
@@ -6,9 +6,18 @@ module Versioning
...
+        #
+        def initialize(prefix: '1')
+          validate_prefix!(prefix)
+          @prefix = prefix.to_s
```
Comment (AliSoftware) [inline lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb:16]: 💡  Could be nice to allow `nil` as an input to the constructor as well, for nicer API 🤔 
```suggestion
          prefix ||= ''
          validate_prefix!(prefix)
          @prefix = prefix.to_s
```

File: lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb (hunk 1 above)
Comment (iangmaia) [reply lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb]: Implemented on 75f41271f15fb3c79b92210d7f32f71102b6ef3c.

File: lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb [hunk 5]
```diff
@@ -19,15 +28,43 @@ class DerivedBuildCodeFormatter
         # @return [String] The formatted build code string.
         #
         def build_code(build_code = nil, version:)
```
Comment (AliSoftware) [inline lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb:31]: Probably not for this PR, but mentioning while I saw it:

Note that this `build_code` parameter is unused, but was initially still kept here so that the common API for all the `*BuildCodeFormatter` have the same API (as mentioned in the YARD comment about it), making it easier to switch from one formatter to another without changing the call sites (abstraction layer pattern).

But in practice I just realized that:
1. We don't have an abstract class for all the `*BuildCodeFormatter` classes like we have for `AbstractVersionFormatter`. In fact, the `*BuildCodeFormatter` classes don't inherit any common parent class, unlike the `*VersionFormatter` classes
2. In practice the API for all `*BuildCodeFormatter` classes is not even consistent (which might actually explain point 1). For example this `DerivedBuildCodeFormatter` as well as `FourPartBuildCodeFormatter` indeed both have a signature of `def build_code(build_code = nil, version:)`… but the `SimpleBuildCodeFormatter` uses `def build_code(version = nil, build_code:)` instead…

That makes me think that (1) either one day we should create an abstract class for the `*BuildCodeFormatter` and unify their API for consistency (2) or we should just abandon the idea of having unified API for all (given the needed input parameters might depend on what the formatter does and needs to compute its value), and remove the unused parameters from the `def build_code(…)` signatures for clarity… 🙃 

---

Also, in any case, I am surprised that `rubocop` didn't complain about it and didn't suggest us to rename this parameter with an underscore to indicate that it being unused is intentional (`def build_code(_build_code = nil, version:)` 🤷

File: spec/derived_build_code_formatter_spec.rb [hunk 6]
```diff
@@ -4,16 +4,102 @@
...
+      end
 
-    it 'derives the build code from version numbers that are two digits' do
-      version = Fastlane::Models::AppVersion.new(12, 34, 56, 78)
-      build_code_string = described_class.new.build_code(version: version)
-      expect(build_code_string.to_s).to eq('112345678')
+      it 'rejects symbols and special characters' do
+        expect { described_class.new(prefix: '@') }.to raise_error(/Prefix must be an integer digit/)
+        expect { described_class.new(prefix: '#') }.to raise_error(/Prefix must be an integer digit/)
+      end
```
Comment (AliSoftware) [inline spec/derived_build_code_formatter_spec.rb:118]: 👌 

(this reminded me of [this meme](https://i.programmerhumor.io/2024/01/programmerhumor-io-testing-memes-programming-memes-d8b14923a839043-758x905.png) 😄 )

File: lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb [hunk 7]
```diff
@@ -6,9 +6,19 @@ module Versioning
       # The `DerivedBuildCodeFormatter` class is a specialized build code formatter for derived build codes.
       # It takes in an AppVersion object and derives a build code from it.
       class DerivedBuildCodeFormatter
+        # Initialize the formatter with a configurable prefix.
+        #
+        # @param [String] prefix The prefix to use for the build code. Must be a single digit (0-9), or the empty string. Defaults to '1' for backward compatibility.
+        #
+        def initialize(prefix: '1')
+          prefix ||= ''
```
Comment (AliSoftware) [inline lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb:14]: Would be nice to add an extra unit test for the `nil` case 🙃

**Review by AliSoftware — APPROVED** [review]
Suggesting one last unit test but otherwise LGTM so approving to unblock.

File: CHANGELOG.md [hunk 8]
```diff
@@ -10,7 +10,7 @@ _None_
 
 ### New Features
 
-_None_
+- Add the possibility to configure DerivedBuildCodeFormatter configure a versioning prefix instead of always defaulting to 1. [#656]
```
Comment (iangmaia) [inline CHANGELOG.md]:
```suggestion
- Add the possibility to configure in DerivedBuildCodeFormatter a versioning prefix instead of always defaulting to 1. [#656]
```

File: lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb (hunk 5 above)
Comment (iangmaia) [reply lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb:31]: > That makes me thing that (1) either one day we should create an abstract class for the *BuildCodeFormatter and unify their API for consistency (2) or we should just abandon the idea of having unified API for all

Oh, agreed.
I quickly saw the parameter and imagined there would be a base, abstract class like the other versioning formatters.

> I am surprised that rubocop didn't complain about it and didn't suggest us to rename this parameter 

True 🤔

File: lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb [hunk 9]
```diff
@@ -6,9 +6,19 @@ module Versioning
       # The `DerivedBuildCodeFormatter` class is a specialized build code formatter for derived build codes.
       # It takes in an AppVersion object and derives a build code from it.
       class DerivedBuildCodeFormatter
+        # Initialize the formatter with a configurable prefix.
+        #
+        # @param [String] prefix The prefix to use for the build code. Must be a single digit (0-9), or the empty string. Defaults to '1' for backward compatibility.
```
Comment (AliSoftware) [inline lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb]: I just realized that funnily enough, I'm not even sure any of our projects use this `DerivedBuildCodeFormatter` yet 🤔 

 - Tumblr uses derived `versionCode` based on `versionName`… but they have their own in-repo implementation (and we've been wanting to migrate to the `release-toolkit` one instead for a while for consistency, but never found the time—see https://linear.app/a8c/issue/AINFRA-155/fix-tumblr-versioning-consistency)
 - I thought for a moment that PocketCasts used that as well because I remember them using some prefix in their `versionCode` to differentiate the Android app, Wear app and Automotive app… but looking at it that's still a `SimpleBuildCodeFormatter` incrementing the versions, and [what I remembered was just an offset between the 3 form factors](https://github.com/Automattic/pocket-casts-android/blob/main/fastlane/Fastfile#L825-L834), not a prefix

… So in practice Gravatar-Android would be the first one to use that class iinm 🙃 

So I wonder if it's really needed to default to `'1'` "for backwards compatibility" given this…
On one hand, better safe that sorry and we can always pass a different value at call site in the product's `Fastfile`. Oh the other hand, if we wanted to change that default prefix, it might actually be now or never 😅

**Review by AliSoftware — APPROVED** [review]

File: lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb (hunk 9 above)
Comment (iangmaia) [reply lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb]: Yeah, that's true. At some point I also realized that Tumblr used their own implementation but wasn't sure about other apps... and anyway this would be a breaking change of the API technically.

But indeed, in practice it would be nicer to not have this default and avoid this issue in the future (especially if more apps adopt this `DerivedBuildCodeFormatter`). Shall we do it? I think I'll do it :feelsgood:

Comment (iangmaia) [reply lib/fastlane/plugin/wpmreleasetoolkit/versioning/formatters/derived_build_code_formatter.rb]: Done in bae9e25c091e5a7904e246faa5bbff4ddee29980

**Review by AliSoftware — APPROVED** [review]
